### PR TITLE
Draft: Resolved 3. Feature: Database Redesign

### PR DESCRIPTION
**Pull request by Ed1196**
1. What additional changes are needed to implement this feature (we want this to be high level but thorough)?

- We are replacing the user1 and user2 ForeignKey's in the Conversation model with a group_users field that is a manytomany relationship to our User model. It will reference all the users that are in a conversation
- Since more than two users can be part of a conversation, all the logic used to post messages and retrieve conversations needs to be changed to handle multiple users.
- We need to change the way we post a message as we now will receive a list of recipient ID's who will receive the message. The front end needs to be able to create a conversation and add a group of users to who the message will be delivered too. 
- Keeping track of the last read message becomes more difficult since we must track the last message read for each individual user. I created a many to many relationship between messages and users so that each message is related to all the users that have read it. So every time a user reads a message, that user gets added to the readBy field in the Message model. At the end we will have a list of users who have read a message for each individual message, making it easier for us to track up to what message each user had read.

2. If our app were already deployed, what steps would we need to take to move to this new feature without disrupting service for current users?

- If our app is already deployed then it is difficult to make these changes as removing or changing fields from Django models can cause a mismatch between code running on the server and the database in use.
- Since we are removing two fields from the conversation model; user1 and user2, if we apply our migrations, those two columns will be gone from the db, but the server still will try to read and write to these fields. We can wrap old fields with deprecate_field() so that the field definition is nullable. Old code that refers to these fields will still work even after the migration are run. 
- We are also adding a new field; group_users, to the conversation fields. This is an issue because there are already objects in our database that were created without this field. We must add a default for this field to the already existing objects and django makes this easy by the use of django-add-default-value. 

The following is a snapshot of what our data could potentially look like once all the changes have been implemented.  

[
    {
        "id": 1,
        "messages": [
            {
                "id": 1,
                "text": "How are you test1",
                "senderId": 1,
                "createdAt": "2021-11-09T16:42:56.562Z",
                "read": false
            },
            {
                "id": 2,
                "text": "How are you test2",
                "senderId": 1,
                "createdAt": "2021-11-09T16:44:47.530Z",
                "read": false
            },
            {
                "id": 3,
                "text": "How are you test3",
                "senderId": 1,
                "createdAt": "2021-11-09T16:44:51.488Z",
                "read": false
            },
            {
                "id": 4,
                "text": "How are you test4",
                "senderId": 1,
                "createdAt": "2021-11-09T16:48:11.542Z",
                "read": false
            }
        ],
        "unread": 0,
        "latestMessageText": "How are you test4",
        "otherUsers": [
            {
                "id": 2,
                "username": "test2",
                "photoUrl": "",
                "online": false
            },
            {
                "id": 3,
                "username": "test3",
                "photoUrl": "",
                "online": false
            }
        ]

### CODE REVIEW

File: server/messenger_backend/models/conversation.py
Comment (andreluciani) [inline server/messenger_backend/models/conversation.py:31]: Instead of using a list of user_ids, can you think of a more efficient way to fetch this data?

File: server/messenger_backend/models/message.py
Comment (andreluciani) [inline server/messenger_backend/models/message.py]: Good, but can you make the read status handing more simple in group conversations?

File: server/messenger_backend/views/api/conversations.py
Comment (andreluciani) [inline server/messenger_backend/views/api/conversations.py]: Good job! But the ticket asks to update just the models, not the API, so this wasn't really needed. :slightly_smiling_face:

File: server/messenger_backend/views/api/messages.py
Comment (andreluciani) [inline server/messenger_backend/views/api/messages.py]: Same as above.

File: server/messenger_backend/views/api/updatemessages.py
Comment (andreluciani) [inline server/messenger_backend/views/api/updatemessages.py]: Same as above :slightly_smiling_face:

**Review by andreluciani — COMMENTED** [review]
Nice job, I really enjoyed reading your proposed database design and migration plan.
I have some more questions, though:

> Since more than two users can be part of a conversation, all the logic used to post messages and retrieve conversations needs to be changed to handle multiple users.
> We need to change the way we post a message as we now will receive a list of recipient ID's who will receive the message. The front end needs to be able to create a conversation and add a group of users to who the message will be delivered too.

- The changes you've mentioned were not very clear. How the front-end code would change to handle group conversations? I would like to know what should be changed on the UI and on the state management.

> Keeping track of the last read message becomes more difficult since we must track the last message read for each individual user. I created a many to many relationship between messages and users so that each message is related to all the users that have read it. So every time a user reads a message, that user gets added to the readBy field in the Message model. At the end we will have a list of users who have read a message for each individual message, making it easier for us to track up to what message each user had read.
- Very nice thing you were concerned about the read status, but using a list/array of users is not very efficient, this is more evident if there is a big group chat and the database has to be updated, for instance. Can you think in a simpler/more efficient way to update this feature?
> Since we are removing two fields from the conversation model; user1 and user2, if we apply our migrations, those two columns will be gone from the db, but the server still will try to read and write to these fields. We can wrap old fields with deprecate_field() so that the field definition is nullable. Old code that refers to these fields will still work even after the migration are run.
- Your migration plan is good, but I miss a few things too: it wasn't very clear if there would be an "intermediate" deploy, and if (and how) the old model would be discarded after deploying the new feature. Can you give more details on that?

> The following is a snapshot of what our data could potentially look like once all the changes have been implemented
- You could have formatted the snapshot better, maybe using the "code" mode ( `<ctrl + e>` shortcut), and where would the data look like this exactly? When fetching conversations? on the Redux state? or it is a general view of how the data would be stored in the database?

Extra question:
- Can you simplify the proposed design if you were not concerned about backward compatibility?

Please append the answers to the questions on the PR description and let us know on Slack when the PR is ready for a second review!

**Comment by Ed1196** [comment]
**The changes you've mentioned were not very clear. How the front-end code would change to handle group conversations? I would like to know what should be changed on the UI and on the state management.**

Front End Changes:

- Our UI needs to have a feature that allows us to create chat rooms and allow us to search for people to add to the chat room. 
- State management for conversations needs to have functions that allow us to add/remove users to these chat rooms
- The field in the conversation **otherUser** needs to be changed to **otherUsers** as we now have multiple users in the conversation
- The **latestReadByOtherId** field that keeps track of the last read message needs to be changed to **latestReadByOthersId** as we need to keep track of the last read message of multiple users. We can use a dictionary object with the key being the user ID and the value being the last read message ID. This way we can access the last read message in O(1) time and its efficient.

**Very nice thing you were concerned about the read status, but using a list/array of users is not very efficient, this is more evident if there is a big group chat and the database has to be updated, for instance. Can you think in a simpler/more efficient way to update this feature?**

- I don’t know if this is a simpler implementation, but it is a more efficient way. I removed the many to many relation from Message to User as too many entries would have been needed. It was not efficient to create a list of users who read a message per message in the conversation, so I created a relation between Conversations and UserMessage; which is a table that holds user_id, latest_message_id and unread_messages; which is the count of unread messages. This way there is only one relation between a conversation and a table that holds the last message read, and the number of unread messages. We would update these as messages get added to the conversation or when a user reads a message. 

The following are tables of how our data will be stored in our database:

This table is of our conversations, we currently have two conversations stored in the DB. 
image

This table below allows us to group the users of different conversations together. Users with ID 1, 2 and 3 are in the same conversation with ID 1 so they get the same lastread_id. 
image

This table below allows us to see the last message read for every user. As we can see user_id 2 is present twice because user with ID 2 is present in two different conversations. 
image

Ultimelty our data will be able to be retrieved and sent to the front end in the following format:
image

**Your migration plan is good, but I miss a few things too: it wasn't very clear if there would be an "intermediate" deploy, and if (and how) the old model would be discarded after deploying the new feature. Can you give more details on that?**

-	The intermediate deployment would involve the following
     o	First we will mark the fields for user1 and user2 as deprecated
      user1 = deprecated_field(models.Foreign….)
         	user2 = deprecated_field(models.Foreign….)
     o	Second we add our new field, but make sure it is nullable
         	group_users = models.ManyToManyField(User, null=True)
     o	By deprecating user1 and user2 makes them become nullable and if new code doesn’t set a value on the field, the code will not crash. This is good since we no longer will use them
     o	By making group_users nullable as well, we will prevent the code from crashing if the value of the field is set to None by servers running old code.
     o	We deploy this intermediate changes

- The final deployment would involve the following
     o	We set our group_users to be not nullable by changing null=False 
        	group_users = models.ManyToManyField(User, null=False)
     o	We also deploy the migration to delete the fields user1 and user2
-  When migrations are not done at the beginning, we must split the changes into two deployments.  

**You could have formatted the snapshot better, maybe using the "code" mode ( <ctrl + e> shortcut), and where would the data look like this exactly? When fetching conversations? on the Redux state? or it is a general view of how the data would be stored in the database?**

What we see below is the new format that the data will be returned to the frontend from the backend. As we can see the latestReadByOtherID holds a dictionary with that key being a users ID and the value being the ID of the last message they read in the conversation. This format will allow the front end to retrieve the ID of the last message to retrieved easily. I provided this snapshot to show that I am not returning an array containing users and their unread messages, but rather a JSON object that can be accessed in O(1). 


**You could have formatted the snapshot better, maybe using the "code" mode ( <ctrl + e> shortcut), and where would the data look like this exactly? When fetching conversations? on the Redux state? or it is a general view of how the data would be stored in the database?**

What we see below is the new format that the data will be returned to the frontend from the backend. As we can see the latestReadByOtherID holds a dictionary with that key being a users ID and the value being the ID of the last message they read in the conversation. This format will allow the front end to retrieve the ID of the last message to retrieved easily. I provided this snapshot to show that I am not returning an array containing users and their unread messages, but rather a JSON object that can be accessed in O(1). 
```
[
    {
        "id": 1,
        "messages": [
            {
                "id": 1,
                "text": "Message 1",
                "senderId": 1,
                "createdAt": "2021-11-12T00:43:57.519Z"
            },
            {
                "id": 2,
                "text": "Message 2",
                "senderId": 1,
                "createdAt": "2021-11-12T00:47:41.985Z"
            },
            {
                "id": 3,
                "text": "Message 3",
                "senderId": 2,
                "createdAt": "2021-11-12T01:19:34.472Z"
            }
        ],
        "unread": 0,
        "latestMessageText": "Message 3",
        "latestReadByOtherId": {
            "2": {
                "latest_message_id": 1
            },
            "3": {
                "latest_message_id": 2
            },
            "1": {
                "latest_message_id": 3
            }
        },
        "otherUsers": [
            {
                "id": 2,
                "username": "test2",
                "photoUrl": "",
                "online": false
            },
            {
                "id": 3,
                "username": "test3",
                "photoUrl": "",
                "online": false
            }
        ]
    }
]
```

image
This function is only used when we don't know the conversation id. I don't know how else we could find a conversation if we don't know the conversation id or use the list of user ids to find the conversation with all these ids.

**Review by FilipeGorodscy — APPROVED** [review]
Awesome and detailed response for the additional questions :)

One little side notes: We wouldn't actually want to store the unread messages count in the database since it is a value that would change very very frequently. We should keep that in the Redux store for better performance and to avoid [flaky tests](https://www.smashingmagazine.com/2021/04/flaky-tests-living-nightmare/) :) 

Approved, but please do not merge since this is only a draft PR.

**Pull request closed without merging**
